zerver/lib/string_validation.py

```diff
@@ -0,0 +1,93 @@
+import unicodedata
+
+from django.utils.translation import gettext as _
+
+from zerver.lib.exceptions import JsonableError
+
+unicode_non_chars = [
+    "\uFDD0",
+    "\uFDD1",
+    "\uFDD2",
+    "\uFDD3",
+    "\uFDD4",
+    "\uFDD5",
+    "\uFDD6",
+    "\uFDD7",
+    "\uFDD8",
+    "\uFDD9",
+    "\uFDDA",
+    "\uFDDB",
+    "\uFDDC",
+    "\uFDDD",
+    "\uFDDE",
+    "\uFDDF",
+    "\uFDE0",
+    "\uFDE1",
+    "\uFDE2",
+    "\uFDE3",
+    "\uFDE4",
+    "\uFDE5",
+    "\uFDE6",
+    "\uFDE7",
+    "\uFDE8",
+    "\uFDE9",
+    "\uFDEA",
+    "\uFDEB",
+    "\uFDEC",
+    "\uFDED",
+    "\uFDEE",
+    "\uFDEF",
+    "\uFFFE",
+    "\uFFFF",
+    "\U0001FFFE",
+    "\U0001FFFF",
+    "\U0002FFFE",
+    "\U0002FFFF",
+    "\U0003FFFE",
+    "\U0003FFFF",
+    "\U0004FFFE",
+    "\U0004FFFF",
+    "\U0005FFFE",
+    "\U0005FFFF",
+    "\U0006FFFE",
+    "\U0006FFFF",
+    "\U0007FFFE",
+    "\U0007FFFF",
+    "\U0008FFFE",
+    "\U0008FFFF",
+    "\U0009FFFE",
+    "\U0009FFFF",
+    "\U0010FFFE",
+    "\U0010FFFF",
+    "\U000AFFFE",
+    "\U000AFFFF",
+    "\U000BFFFE",
+    "\U000BFFFF",
+    "\U000CFFFE",
+    "\U000CFFFF",
+    "\U000DFFFE",
+    "\U000DFFFF",
+    "\U000EFFFE",
+    "\U000EFFFF",
+    "\U000FFFFE",
+    "\U000FFFFF",
+]
+
+
+def check_invalid_string(var: str, entity: str, max_length: int) -> None:
+    if var.strip() == "":
+        raise JsonableError(_("{} can't be empty!").format(entity))
```

Comment:
This function name feels really not-right.  "valid" is one of those adjectives which is really squirrelly -- it doesn't tell me much information unless it's universally clear what a valid or invalid thing of its class is.  For example, "valid number" is pretty ambiguous, since that could potentially be integers, floats, complex numbers, etc.  "valid string" is similarly pretty unclear -- valid for _what_?  When should I reach for this function?

There's also another smell here that we're barking up the wrong tree doing it this way -- there's no way to translate this error.  For example, French translators will translate "{} can't be empty!" as "{} ne peut pas être vide!" which, if thrown, will produce the polyglot "Stream name ne peut pas être vide!" and not "Le nom de canal ne peut pas être vide!"

I think this is an example of [the wrong abstraction](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction); both the inability to have a good name for it, and the inability to return sensible error messages, point that we're trying too hard to reduce duplication, and it's making it _harder_ to write clear code.

I think the right thing here is to have a function for stream names, and a function for instance names.  Put them in the same file so they can both use `unicode_non_chars`, and so their parallelism is evident, but making them their own functions means their errors can be very straightforward, they won't have extraneous flags, and it'll be clear when to call them.

If we really want to de-duplicate something, I think the only part of this that we could pull out is a predicate (function that returns a `bool`) that checks if a character is in `Cc`, `Cs`, or a non-character.